### PR TITLE
Add map of virtual memory and opencl memory object in opencl branch. 

### PR DESCRIPTION
**Pull request by wujunkai166**
Adopt the map idea in clcaffe branch. After applying this approach, function parameters with pattern (cl_mem A, int offA) can be simplified to (Dtype* A), which behaves just like functions using CUDA. This approach can significantly reduce the difference between opencl path and CUDA path in code level and make it easier to merge the changes from BVLC master branch.

Moreover, adding this mapping technique doesn't bring any overhead to the performance of the code.

### CODE REVIEW

**Comment by wujunkai166** [comment]
@gongzg

**Comment by gongzg** [comment]
@wujunkai166 

There are some conflicts with the latest opencl branch, please rebase this PR. Thanks.

**Comment by wujunkai166** [comment]
@gongzg 

I've solved the conflicts and push to the opencl branch. Please check again. Thanks.

File: include/caffe/greentea/greentea_math_functions.hpp [hunk 1]
```diff
@@ -22,140 +22,89 @@
...
-void greentea_gpu_memcpy(const uint_tp N, const cl_mem X, const int_tp offX,
-                         void *Y, viennacl::ocl::context *ctx);
-
-void greentea_gpu_memcpy(const uint_tp N, const void* X, cl_mem Y,
-                         const int_tp offY, viennacl::ocl::context *ctx);
-
-void greentea_gpu_memcpy(const uint_tp N, const cl_mem X, const int_tp offX,
-                         cl_mem Y, const int_tp offY,
-                         viennacl::ocl::context *ctx);
+void caffe_gpu_memset(const uint_tp N, const int_tp alpha, void* X);
```
Comment (gongzg) [inline include/caffe/greentea/greentea_math_functions.hpp]: We should use those prototypes in math_functions.hpp directly and there is no need to have a greentea_math_functions.hpp any more. This is one of the reason why we introduce this mechanism.

File: include/caffe/util/cl_state.hpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+namespace caffe {
+
+template<typename T>
+struct ClMemOff {
+  cl_mem memobj;
+  size_t offset;  // offset in elements
+};
+
+class ClState {
```
Comment (gongzg) [inline include/caffe/util/cl_state.hpp]: Please add some comments here to explain the purpose we introduce clstate here brieflly.

File: include/caffe/util/math_functions.hpp [hunk 3]
```diff
@@ -35,6 +35,8 @@ void caffe_cpu_axpby(const int_tp N, const Dtype alpha, const Dtype* X,
 template<typename Dtype>
 void caffe_cpu_copy(const int_tp N, const Dtype* X, Dtype* Y);
 
+void caffe_copy(const int_tp N, const void* X, void* Y);
```
Comment (gongzg) [inline include/caffe/util/math_functions.hpp]: why add a caffe_copy prototype here? There is already one template in the same file.

File: src/caffe/common.cpp [hunk 4]
```diff
@@ -497,13 +501,15 @@ void Caffe::SetDevice(const int device_id) {
 #if defined(USE_GREENTEA) && defined(USE_FFT)
   Get().cl_fft_state_.setup();
 #endif
+
```
Comment (gongzg) [inline src/caffe/common.cpp]: useless blank line, should remove it.

File: src/caffe/common.cpp [hunk 5]
```diff
@@ -497,13 +501,15 @@ void Caffe::SetDevice(const int device_id) {
...
 }
 
 // Should call explicitly for OCL + FFT
 void Caffe::TeardownDevice(const int device_id) {
 #if defined(USE_GREENTEA) &&defined(USE_FFT)
   Get().cl_fft_state_.teardown();
 #endif
+
```
Comment (gongzg) [inline src/caffe/common.cpp]: Ditto.

File: src/caffe/common.cpp [hunk 6]
```diff
@@ -27,6 +27,10 @@ namespace caffe {
 // Make sure each thread can have different values.
 static boost::thread_specific_ptr<Caffe> thread_instance_;
 
+#ifdef USE_GREENTEA
+ClState Caffe::cl_state_;
```
Comment (gongzg) [inline src/caffe/common.cpp]: So the cl_state_ is a global object here, you need to consider the thread safe issue in cl_state.cpp.

**Comment by gongzg** [comment]
@wujunkai166 I added some inline comments for your commits. Besides that, I found there are many format issue, you can simply check them by git show --check. Please fix them before submit the next PR.  And after merge this PR, it breaks the build as below, please check and fix them. Thanks.

In file included from /caffe/include/caffe/common.hpp:25:0,
                 from /caffe/src/caffe/greentea/greentea_math_functions.cpp:8:
/caffe/src/caffe/greentea/greentea_math_functions.cpp: In function ‘void caffe::caffe_gpu_axpy(int32_t, Dtype, const Dtype*, Dtype*) [with Dtype = float; int32_t = int]’:
/caffe/src/caffe/greentea/greentea_math_functions.cpp:638:31: error: ‘MemX’ was not declared in this scope
         clblasSaxpy(N, alpha, MemX, offX,
                               ^
/caffe/include/caffe/greentea/greentea.hpp:79:28: note: in definition of macro ‘GREENTEA_CL_BLAS_CHECK’
     {clblasStatus status = condition; \
                            ^
/caffe/src/caffe/greentea/greentea_math_functions.cpp: In function ‘void caffe::caffe_gpu_axpy(int32_t, Dtype, const Dtype*, Dtype*) [with Dtype = double; int32_t = int]’:
/caffe/src/caffe/greentea/greentea_math_functions.cpp:712:31: error: ‘MemX’ was not declared in this scope
         clblasDaxpy(N, alpha, MemX, offX,
                               ^
/caffe/include/caffe/greentea/greentea.hpp:79:28: note: in definition of macro ‘GREENTEA_CL_BLAS_CHECK’
     {clblasStatus status = condition; \
                            ^
/caffe/src/caffe/greentea/greentea_math_functions.cpp: In function ‘void caffe::caffe_gpu_asum(int32_t, const Dtype*, Dtype*) [with Dtype = float; int32_t = int]’:
/caffe/src/caffe/greentea/greentea_math_functions.cpp:1147:24: error: request for member ‘memobj’ in ‘memScratch’, which is of non-class type ‘float*’
             memScratch.memobj, 1, &queue, 0, NULL, NULL));
                        ^
/caffe/include/caffe/greentea/greentea.hpp:79:28: note: in definition of macro ‘GREENTEA_CL_BLAS_CHECK’
     {clblasStatus status = condition; \
                            ^
/caffe/src/caffe/greentea/greentea_math_functions.cpp: In function ‘void caffe::caffe_gpu_asum(int32_t, const Dtype*, Dtype*) [with Dtype = double; int32_t = int]’:
/caffe/src/caffe/greentea/greentea_math_functions.cpp:1227:56: error: cannot convert ‘double*’ to ‘float*’ in initialization
    CL_MEM_READ_WRITE, n * sizeof(cl_double), NULL, err));
                                                        ^
/caffe/src/caffe/greentea/greentea_math_functions.cpp:1229:54: error: conversion from ‘caffe::ClMemOff<float>’ to non-scalar type ‘caffe::ClMemOff<double>’ requested
  ClMemOff<double> bufScratch = clState.get_buffer_mem(memScratch);
                                                      ^
In file included from /caffe/include/caffe/common.hpp:25:0,
                 from caffe/src/caffe/greentea/greentea_math_functions.cpp:8:
/caffe/src/caffe/greentea/greentea_math_functions.cpp:1233:24: error: request for member ‘memobj’ in ‘memScratch’, which is of non-class type ‘float*’
             memScratch.memobj, 1, &queue, 0, NULL, NULL));
                        ^

**Comment by wujunkai166** [comment]
@gongzg I solved all of the bugs and deleted two head files (greentea_match_functions.hpp and greentea_im2col.hpp). Still work on format issues since there are quite a lot places need to be modified.

**Comment by gongzg** [comment]
@wujunkai166 please also fix the lint errors which you could check easily with make lint. thanks.

**Comment by wujunkai166** [comment]
@gongzg Format and lint errors are solved.

**Comment by gongzg** [comment]
@wujunkai166 I found this PR cause some regressions in the test suite.Please double check, thx.

**Comment by wujunkai166** [comment]
@gongzg Test regressions are solved. Please review and test it. Thanks.

**Comment by gongzg** [comment]
@wujunkai166 I just merged this PR to opencl-virtptr branch and submit a PR to the upstream opencl branch. So I'm going to close this PR and let's discuss at the https://github.com/BVLC/caffe/pull/4421. Thanks.

**Pull request closed without merging**